### ISSUE TITLE
Text mining

### ISSUE THREAD

**Issue by katopaulas** [rflamary/POT#30]
Hello there,

i am currently toying around POT comparing texts. I have a dictionary of 46k terms and im trying to compare 120k documents. Every document has at most 10-15 words(bibtex titles) so comparing 2 distributions - texts will result in comparing 2 [46000,1] vectors with at most 10 non zero entries.

Are there any suggestions in the process because the naive approach is too slow. Comparing 10k documents takes 2 days.
( emd2(p,q,C) p,q are [46k,1] C is [46k,46k]
Sinkhorn is even slower!

Thanks in advance!

**Comment by rflamary** [comment]
Helo @katopaulas,

If you have only 10 nonzeros enries, you should solve the subpproblem only with the active words.

we have a function that clean the problem here http://pot.readthedocs.io/en/stable/all.html#ot.utils.clean_zeros

but I would recommend to do it by hand 

```python
# p vector, q matrix
p2=p[p>0]
q2=q[q.sum(1)>0,:]

C2=C[p>0,:][:,q.sum(1)>0].copy()

emd2(p2,q2,C2)

```

also note that while the solver in emd and emd2 does not work using multi-processors, emd2 can compute multiple emd in parallel (beware of the memory cost, use cleaned C matrices).

Another approach would be to do a word clustering and compute OT on the smaller histograms to perform a coarse search and solve the high resolution problems as a second step.

**Comment by katopaulas** [comment]
Hello again, super fast answer!

I don't want to compute emd with vector / matrix . I will end up doing computations that are already done in previous iterations(i want only the upper triangular of the EMD of the distances between docs).

Does the same method apply if the vectors have different amount of non zero entries?
The dimensions of the Cost matrix after the removal of zeros confuse me a bit.
I will try it and inform you!

And again thank you!!!

**Comment by remuspetrupop** [comment]
Hello,

I am trying to use the sinkhorn algorithm to calculate the distance between two images. First, I am transforming the images into two probability vectors, which end up containing many values of zero. Then I am getting the following error:

Warning: numerical errors at iteration 0
when calling
d_sinkhorn = ot.sinkhorn2(v1, v2, cm, reg)

It still outputs a result for the distance, however, I was wondering whether I should clean the vectors first from the non-zero entries. Also, adding a constant value to the vectors before normalizing stops the error (since all entries are non-zero now) but I am not sure whether this approach is correct. In particular, my goal is to compare a bunch of images and select the top ones with largest distances.

Many thanks in advance!

**Comment by rflamary** [comment]
Hello,

The original question has been answered, i'm closing the issue.
